### PR TITLE
3.0 fixing belongstomany id joindata

### PR DESCRIPTION
**Pull request by mylux**
Added a bugfix. It wasn't possible to create a new entity in the following conditions:

Article->BelongsToMany Tags.

When the article is new, but the tags are already existent. When creating a new entity, while passing id and _joinData for each tag, instead of search for the tag with the defined id, it was adding a new entity (with new=>true) and with the field id filled.

"id" field is not hard coded. It is set to primary key field and it should support multiple keys constituting the primary key.

With this correction, it should look for each tag with the defined id and add the _joinData accordingly and only add a new Tag entity if one with this id is not found in the database.
An example can be found in the test named: testOneBelongsToManyJoinDataAssociatedWithIds

### CODE REVIEW

**Comment by markstory** [comment]
Looks [like you have a few PHPCS](https://travis-ci.org/cakephp/cakephp/jobs/54510902) errors.

File: src/ORM/Marshaller.php [hunk 1]
```diff
@@ -105,6 +105,18 @@ public function one(array $data, array $options = [])
         $propertyMap = $this->_buildPropertyMap($options);
 
         $schema = $this->_table->schema();
+        $primaryKey = $schema->primaryKey();
+
+        if ( array_intersect($primaryKey, array_keys($data)) == $primaryKey ) {
```
Comment (markstory) [inline src/ORM/Marshaller.php]: Won't this always pass when the primary key is present? Also what about associated values that should be marshaled under this entity?

File: tests/TestCase/ORM/MarshallerTest.php [hunk 2]
```diff
@@ -495,6 +495,52 @@ public function testOneBelongsToManyJoinDataAssociated()
...
+        );
+        $this->assertInstanceOf(
+            'Cake\ORM\Entity',
+            $result->tags[1]
+        );
+
+        $this->assertEquals(false,$result->tags[0]->isNew() );
+        $this->assertEquals(false,$result->tags[1]->isNew() );
+        $this->assertEquals( TableRegistry::get('tags')->get(1)->tag,$result->tags[0]->tag );
+        $this->assertEquals( TableRegistry::get('tags')->get(2)->tag,$result->tags[1]->tag );
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: You should also have asserts for `_joinData` contents, and the users inside them.

File: tests/TestCase/ORM/MarshallerTest.php [hunk 3]
```diff
@@ -495,6 +495,51 @@ public function testOneBelongsToManyJoinDataAssociated()
...
+        );
+        $this->assertInstanceOf(
+            'Cake\ORM\Entity',
+            $result->tags[1]
+        );
+
+        $this->assertEquals(false, $result->tags[0]->isNew());
+        $this->assertEquals(false, $result->tags[1]->isNew());
+        $this->assertEquals(TableRegistry::get('tags')->get(1)->tag, $result->tags[0]->tag);
+        $this->assertEquals(TableRegistry::get('tags')->get(2)->tag, $result->tags[1]->tag);
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: You should verify the nested `_joinData` and `_joinData.user` associations were marshaled correctly as well.

File: src/ORM/Marshaller.php (hunk 1 above)
Comment (mylux) [inline src/ORM/Marshaller.php]: The idea of this condition is exactly to check if the primary key is present. This condition is made to support a primary key wich is composed by 1 or more fields. I don't get exactly what you said about associated values, kindly ask you to explain a bit more what you mean. Thanks!

File: tests/TestCase/ORM/MarshallerTest.php (hunk 3 above)
Comment (mylux) [inline tests/TestCase/ORM/MarshallerTest.php]: I just improved the Test case for that. Thank you!

File: src/ORM/Marshaller.php [hunk 4]
```diff
@@ -105,6 +105,18 @@ public function one(array $data, array $options = [])
         $propertyMap = $this->_buildPropertyMap($options);
 
         $schema = $this->_table->schema();
+        $primaryKey = $schema->primaryKey();
+
+        if (array_intersect($primaryKey, array_keys($data)) == $primaryKey) {
+            $record = $this->_table->find('all');
+            foreach ($primaryKey as $pkey) {
+                $record->where(["$pkey" => $data[$pkey] ]);
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: It should have the table name as prefix in the column to avoid collisions.

File: src/ORM/Marshaller.php [hunk 5]
```diff
@@ -105,6 +105,18 @@ public function one(array $data, array $options = [])
...
+            }
+            if ($record->count() > 0) {
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: I would try to avoid this count and just return `first()` if any

Comment (ionas) [inline src/ORM/Marshaller.php]: $record->where(["$pkey" => $data[$pkey] ]);
... what are the quotes for?

File: src/ORM/Marshaller.php (hunk 4 above)
Comment (mylux) [inline src/ORM/Marshaller.php]: Hi Lorenzo. Thanks for your comment. Honestly I don't see the possibility of collisions once I am not joining tables. But if you still think it is necessary to add table names I can do it without problem, I just kindly ask you to confirm.

File: tests/TestCase/ORM/MarshallerTest.php (hunk 2 above)
Comment (mylux) [inline tests/TestCase/ORM/MarshallerTest.php]: Thanks for your suggestion. It was added.

File: src/ORM/Marshaller.php (hunk 5 above)
Comment (mylux) [inline src/ORM/Marshaller.php]: Hi Ionas, my idea was to keep clear that we are dealing with strings.
Hi Lorenzo, thanks again for your suggestion. In my last commit I did this.

File: src/ORM/Marshaller.php (hunk 4 above)
Comment (markstory) [inline src/ORM/Marshaller.php]: You're not adding joins, but someone somewhere will use callbacks or override find() to add joins. We've had a number of similar issues reported when we thought we were 'safe'.

Comment (mylux) [inline src/ORM/Marshaller.php]: Alright, them. If that's the case, I will add the table names. Thanks for your help.

File: src/ORM/Marshaller.php [hunk 6]
```diff
@@ -105,6 +105,22 @@ public function one(array $data, array $options = [])
         $propertyMap = $this->_buildPropertyMap($options);
 
         $schema = $this->_table->schema();
+        $primaryKey = $schema->primaryKey();
+
+        if (array_intersect($primaryKey, array_keys($data)) == $primaryKey) {
+            $tableName = $this->_table->table();
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: it should be `->alias()` instead of `table()`

File: src/ORM/Marshaller.php [hunk 7]
```diff
@@ -105,6 +105,22 @@ public function one(array $data, array $options = [])
...
+            $record = $this->_table->find('all');
+            foreach ($primaryKey as $pkey) {
+                $record->where(["$tableName.$pkey" => $data[$pkey] ]);
```
Comment (lorenzo) [inline src/ORM/Marshaller.php]: There is an extra space after `]`

Comment (mylux) [inline src/ORM/Marshaller.php]: What surprised me is that phpcs tests made locally didn't pointed this. I just fixed this.
Thank you

File: src/ORM/Marshaller.php (hunk 6 above)
Comment (mylux) [inline src/ORM/Marshaller.php]: Fixed that. Thanks.

File: tests/TestCase/ORM/MarshallerTest.php [hunk 8]
```diff
@@ -495,6 +495,68 @@ public function testOneBelongsToManyJoinDataAssociated()
...
+            'Cake\ORM\Entity',
+            $result->tags[0]->_joinData->user
+        );
+
+        $this->assertInstanceOf(
+            'Cake\ORM\Entity',
+            $result->tags[1]->_joinData->user
+        );
+        $this->assertEquals(false, $result->tags[0]->isNew());
+        $this->assertEquals(false, $result->tags[1]->isNew());
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: assertFalse() instead.

**Comment by lorenzo** [comment]
I can take care of fixing the rest of the comments. Thanks so much for bearing with us @mylux and for this fix!

File: src/ORM/Marshaller.php [hunk 9]
```diff
@@ -105,6 +105,22 @@ public function one(array $data, array $options = [])
         $propertyMap = $this->_buildPropertyMap($options);
 
         $schema = $this->_table->schema();
+        $primaryKey = $schema->primaryKey();
+
+        if (array_intersect($primaryKey, array_keys($data)) == $primaryKey) {
+            $tableName = $this->_table->alias();
+            $record = $this->_table->find('all');
+            foreach ($primaryKey as $pkey) {
+                $record->where(["$tableName.$pkey" => $data[$pkey]]);
```
Comment (ADmad) [inline src/ORM/Marshaller.php]: Prefixing field with table name/alias similar to `"$tableName.$pkey"` will be often needed by users. It be nice to have `Table::escapeField()` similar to `Model::escapeField()` in 2.x.

Comment (kamui545) [inline src/ORM/Marshaller.php]: :+1: For now I have an "AppTable" with this function cause I often need it for my custom finders.

Comment (markstory) [inline src/ORM/Marshaller.php]: I wouldn't call it escapeField though as that is pretty misleading. Instead we could call it aliasField().

Comment (ADmad) [inline src/ORM/Marshaller.php]: :+1: to better name.

**Pull request merged**